### ISSUE TITLE
False positive on `non_exhaustive_omitted_patterns_lint` with closures

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust#99815]
I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5eda6149e4d8df5ec233e5133d06b0f1)):

```rust
#![warn(non_exhaustive_omitted_patterns)]

// in other library:
// #[non_exhaustive] pub enum Foo { A, B }
use lib::Foo;

fn foo(_: impl Fn(Foo)) {}

fn main() {
    foo(|_: Foo| { todo!() });
}
```

I expected to see this happen: no warnings.

Instead, this happened:
```
warning: some variants are not matched explicitly
  --> src/main.rs:10:14
   |
10 |         foo(|_: Foo| { todo!() });
   |              ^ patterns `A` and `B` not covered
   |
```

### Meta
```
1.64.0-nightly (2022-07-26 4d6d601c8a83284d6b23)
```

cc #89554

**Comment by Nemo157** [comment]
(Note: it's not to do with the `_` pattern, in the real code I found this in the argument was used).

**Comment by Mark-Simulacrum** [comment]
I think this is a good thing to fix, but I don't think this would block stabilization since:

* It's unlikely you're enabling the lint across an entire crate -- most of the time, it'll be limited to a few tests or specific matches, probably.
* The lint can be fixed in the future (we should definitely do that though!)
